src/tiled/objectselectiontool.cpp

```diff
@@ -595,24 +593,25 @@ void ObjectSelectionTool::mouseReleased(QGraphicsSceneMouseEvent *event)
             auto underlyingObjects = objectItemsAt(event->scenePos());
             if (underlyingObjects.isEmpty())
                 break;
-            int currRotationIndex = underlyingObjects.indexOf(mSelectedInUnderlyingRotation);
-            mClickedObjectItem = mSelectedInUnderlyingRotation
-                    = underlyingObjects.at((currRotationIndex + 1) % underlyingObjects.size());
+
+            int lastSelectedIndex = -1;
+            for (auto selected : selection)
+                lastSelectedIndex = std::max(lastSelectedIndex, underlyingObjects.indexOf(selected));
+            do lastSelectedIndex = (lastSelectedIndex + 1) % underlyingObjects.size();
+            while (selection.contains(underlyingObjects.at(lastSelectedIndex))
+                   && lastSelectedIndex != underlyingObjects.size() - 1);
+            mClickedObjectItem = underlyingObjects.at(lastSelectedIndex);
```

Comment:
It doesn't support a last->first iteration. If the bottom part of the stack is Shift-selected, your code always point at the topmost element.